include/net/mqtt.h

```diff
@@ -1,474 +1,685 @@
+	/** Internal. Mutex to protect access to the client instance. */
+	struct k_mutex mutex;
+
+	/** Internal. Wall clock value (in milliseconds) of the last activity
+	 *  that occurred. Needed for periodic PING.
+	 */
+	u32_t last_activity;
+
+	/** Internal. Client's state in the connection. */
+	u32_t state;
+
+	/** Internal.  Packet length read so far. */
+	u32_t rx_buf_datalen;
+
+	/** Internal. Remaining payload length to read. */
+	u32_t remaining_payload;
+};
 
 /**
- * Sends the MQTT PUBACK message with the given packet id
- *
- * @param [in] ctx MQTT context structure
- * @param [in] id MQTT Packet Identifier
- *
- * @retval 0 on success
- * @retval -EINVAL
- * @retval -ENOMEM
- * @retval -EIO
+ * @brief MQTT Client definition to maintain information relevant to the
+ *        client.
  */
-int mqtt_tx_puback(struct mqtt_ctx *ctx, u16_t id);
+struct mqtt_client {
+	/** MQTT client internal state. */
+	struct mqtt_internal internal;
+
+	/** MQTT transport configuration and data. */
+	struct mqtt_transport transport;
+
+	/** Unique client identification to be used for the connection. */
+	struct mqtt_utf8 client_id;
+
+	/** Broker details, for example, address, port. Address type should
+	 *  be compatible with transport used.
+	 */
+	const void *broker;
+
+	/** User name (if any) to be used for the connection. NULL indicates
+	 *  no user name.
+	 */
+	struct mqtt_utf8 *user_name;
+
+	/** Password (if any) to be used for the connection. Note that if
+	 *  password is provided, user name shall also be provided. NULL
+	 *  indicates no password.
+	 */
+	struct mqtt_utf8 *password;
+
+	/** Will topic and QoS. Can be NULL. */
+	struct mqtt_topic *will_topic;
+
+	/** Will message. Can be NULL. Non NULL value valid only if will topic
+	 *  is not NULL.
+	 */
+	struct mqtt_utf8 *will_message;
+
+	/** Application callback registered with the module to get MQTT events.
+	 */
+	mqtt_evt_cb_t evt_cb;
+
+	/** Receive buffer used for MQTT packet reception in RX path. */
+	u8_t *rx_buf;
+
+	/** Size of receive buffer. */
+	u32_t rx_buf_size;
+
+	/** Transmit buffer used for creating MQTT packet in TX path. */
+	u8_t *tx_buf;
+
+	/** Size of transmit buffer. */
+	u32_t tx_buf_size;
+
+	/** MQTT protocol version. */
+	u8_t protocol_version;
+
+	/** Will retain flag, 1 if will message shall be retained persistently.
+	 */
+	u8_t will_retain : 1;
+
+	/** Clean session flag indicating a fresh (1) or a retained session (0).
+	 *  Default is 1.
+	 */
+	u8_t clean_session : 1;
+};
 
 /**
- * Sends the MQTT PUBCOMP message with the given packet id
+ * @brief Initializes the client instance.
  *
- * @param [in] ctx MQTT context structure
- * @param [in] id MQTT Packet Identifier
+ * @param[in] client Client instance for which the procedure is requested.
+ *                   Shall not be NULL.
  *
- * @retval 0 on success
- * @retval -EINVAL
- * @retval -ENOMEM
- * @retval -EIO
+ * @note Shall be called before connecting the client in order to avoid
```

Comment:
Please mention that this should be called even before setting other fields in the client. I may suggest: "Shall be called to initialize client structure, before setting any client parameters and before connecting to broker."